torchaudio/functional/functional.py

```diff
@@ -1631,3 +1637,274 @@ def rnnt_loss(
         return costs.sum()
 
     return costs
+
+
+def _compute_mat_trace(input: torch.Tensor, dim1: int = -1, dim2: int = -2) -> torch.Tensor:
+    r"""Compute the trace of a Tensor along ``dim1`` and ``dim2`` dimensions.
+
+    Args:
+        input (torch.Tensor): Tensor of dimension `(..., channel, channel)`
+        dim1 (int, optional): the first dimension of the diagonal matrix
+            (Default: -1)
+        dim2 (int, optional): the second dimension of the diagonal matrix
+            (Default: -2)
+
+    Returns:
+        torch.Tensor: trace of the input Tensor
+    """
+    assert input.ndim >= 2, "The dimension of the tensor must be at least 2."
+    assert input.shape[dim1] == input.shape[dim2], "The size of ``dim1`` and ``dim2`` must be the same."
+    input = torch.diagonal(input, 0, dim1=dim1, dim2=dim2)
+    return input.sum(dim=-1)
+
+
+def _tik_reg(mat: torch.Tensor, reg: float = 1e-7, eps: float = 1e-8) -> torch.Tensor:
+    """Perform Tikhonov regularization (only modifying real part).
+    Args:
+        mat (torch.Tensor): input matrix (..., channel, channel)
+        reg (float, optional): regularization factor (Default: 1e-8)
+        eps (float, optional): a value to avoid the correlation matrix is all-zero (Default: ``1e-8``)
+
+    Returns:
+        torch.Tensor: regularized matrix (..., channel, channel)
+    """
+    # Add eps
+    C = mat.size(-1)
+    eye = torch.eye(C, dtype=mat.dtype, device=mat.device)
+    with torch.no_grad():
+        epsilon = _compute_mat_trace(mat).real[..., None, None] * reg
+        # in case that correlation_matrix is all-zero
+        epsilon = epsilon + eps
+    mat = mat + epsilon * eye[..., :, :]
+    return mat
+
+
+def compute_power_spectral_density_matrix(
+    specgram: Tensor,
+    mask: Optional[Tensor] = None,
+    multi_mask: bool = False,
+    normalize: bool = True,
+    eps: float = 1e-10,
+) -> Tensor:
+    """Compute cross-channel power spectral density (PSD) matrix.
+
+    Args:
+        specgram (Tensor): multi-channel complex-valued STFT matrix.
+            Tensor of dimension `(..., channel, freq, time)`
```

Comment:
The shape is to align with other methods in torchaudio: https://github.com/pytorch/audio/blob/main/CONTRIBUTING.md?plain=1#L226

But your point makes sense, for the solution, we can keep the argument ``(..., channel, freq, time)`` and do transpose internally to reduce memory.

Reply:
But I doubt why the speed is faster. I thought tensor operation at one time should be faster than for loop. Or is the for loop in parallel?

Reply:
It can be faster. 
With batched operations you get the advantage, that the python overhead is smaller and the C or C++ code can use vector instructions. But at some point, the python overhead is neglectable and vector instructions cannot further increase the execution time. On the other side, when your data gets too large, it will no longer fit into the L1/L2/L3 cache. So this will slow down your program, because the data between each operation has to be moved between the caches and maybe also the RAM.

In WPE it turned out, that a loop over the frequency is sometimes faster than the batched implementation, especially for longer audio clips. (On the GPU this is different, but I got only a speedup of 3 with CUPY vs CPU and profiled it only once)

Reply:
> The shape is to align with other methods in torchaudio:

I expected that answer. We gave that idea up, because different algorithms/approaches prefer different shapes and some of them get very ugly or slow, when you enforce a particular shape.

I just wanted to mention this, and aligning shapes is a very good argument.